### PR TITLE
Add normal refinement via minimum spanning tree propagation

### PR DESCRIPTION
**Pull request by chambbj**
Adds switch (default ~`false`~ `true`) to refine the normal orientation by implementing the method of minimum spanning tree normal propagation.

### CODE REVIEW

**Comment by abellgithub** [comment]
IMO most of this doesn't belong here.  It belongs in some utility-type file that exposes this via some function that constructs a minimal spanning tree.

Also, it's not clear to me what this is actually doing and why you'd want this.  It needs documentation/reference.

It also needs tests.

The `refine` default is `true` - the above comment says `false`.

I can look at this in a more detailed way when I understand it.

**Comment by chambbj** [comment]
Please see "Hoppe, Hugues, et al. Surface reconstruction from unorganized points. Vol. 26. No. 2. ACM, 1992" for more on the method, specifically section 3.3.

The need arises because there can be ambiguity in the orientation of the approximating planes for the normals. We already offer the ability to flip towards a viewpoint/positive Z for the same basic reason. This is yet another method of making them more consistent across the dataset.

I can get you some test and doc updates.

I think if you wanted to abstract any of the MST stuff, you could do that after the fact. This is a bit specialized, in that it is constructed on the fly (the propagation part), so I don't know how well it separates, but am open to the idea in general.

File: filters/NormalFilter.cpp [hunk 1]
```diff
@@ -112,53 +138,177 @@ void NormalFilter::prepared(PointTableRef table)
             << "Viewpoint provided. Ignoring always_up = TRUE." << std::endl;
         m_args->m_up = false;
     }
+
+    // The query point is returned as a neighbor of itself, so we must increase
+    // k by one to get the desired number of neighbors.
+    ++m_args->m_knn;
```
[the remark this answers is not in the document]
Comment (chambbj) [reply filters/NormalFilter.cpp]: Yeah, but I'd argue 1) it should've been this way all along (here and plenty of other places in the code) and 2) don't know that the sensitivity between `k` and `k+1` would be all that detectable either way.

File: filters/NormalFilter.cpp [hunk 2]
```diff
@@ -112,53 +138,177 @@ void NormalFilter::prepared(PointTableRef table)
...
+            // viewpoint by taking the dot product of the vector connecting the
+            // point with the viewpoint and the normal. Flip the normal, where
+            // the dot product is negative.
+            float dx = static_cast<float>(m_args->m_viewpoint.x() -
+                                          p.getFieldAs<double>(Id::X));
+            float dy = static_cast<float>(m_args->m_viewpoint.y() -
+                                          p.getFieldAs<double>(Id::Y));
+            float dz = static_cast<float>(m_args->m_viewpoint.z() -
+                                          p.getFieldAs<double>(Id::Z));
+            Vector3d vp(dx, dy, dz);
```
[the remark this answers is not in the document]
Comment (chambbj) [reply filters/NormalFilter.cpp]: Agreed, it's a holdover from the original codebase that I did not catch.

File: filters/NormalFilter.cpp [hunk 3]
```diff
@@ -112,53 +138,177 @@ void NormalFilter::prepared(PointTableRef table)
...
+        p.setField(Id::NormalY, normal[1]);
+        p.setField(Id::NormalZ, normal[2]);
+        p.setField(Id::Curvature, curvature);
+    }
+}
 
-        double sum = eval[0] + eval[1] + eval[2];
-        view.setField(Dimension::Id::Curvature, i,
-                      sum ? std::fabs(eval[0] / sum) : 0);
+void NormalFilter::refine(PointView& view, KD3Index& kdi)
```
[the remark this answers is not in the document]
Comment (chambbj) [reply filters/NormalFilter.cpp]: Yes, I fell short of choosing the starting point that way. Shouldn't be much work IIRC.

File: filters/NormalFilter.cpp [hunk 4]
```diff
@@ -112,53 +138,177 @@ void NormalFilter::prepared(PointTableRef table)
...
+            }
+            else if (!inMST[edge.m_v1])
+            {
+                newIdx = edge.m_v1;
+                normal = N2;
+            }
+            else
+                continue;
+
+            // Where the dot product of the normals is less than 0, invert the
```
[the remark this answers is not in the document]
Comment (chambbj) [reply filters/NormalFilter.cpp]: Yes, just sticking with the as published algorithm for now.

File: filters/NormalFilter.cpp [hunk 5]
```diff
@@ -112,53 +138,177 @@ void NormalFilter::prepared(PointTableRef table)
...
+    while (count < view.size())
+    {
+        // Find the PointId of the next point not currently part of the minimum
+        // spanning tree.
+        while (inMST[nextIdx])
+            ++nextIdx;
+
+        // Lambda to add the current PointId to the minimum spanning tree and
+        // update the edge queue.
+        auto update = [&](PointId updateIdx) {
```
[the remark this answers is not in the document]
Comment (chambbj) [reply filters/NormalFilter.cpp]: I'll break this out into it's own function. That'll simplify `refine()` to address your other comment anyway.

File: filters/NormalFilter.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (abellgithub) [reply filters/NormalFilter.cpp]: I don't disagree.  Was just pointing it out.  I think it's fine.

File: filters/NormalFilter.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (chambbj) [reply filters/NormalFilter.cpp]: I'm going to leave this as-is for now. We can improve it later.

**Comment by xiaoxiaozhu007** [comment]
Hi, would you please check if the refine function still available for filters.normal? 
I got the error "RuntimeError: filters.normal: Unexpected argument 'refine'." when I tried to enable the refine function. I don't know the cause of this error. 
Also, if the function is enabled automatically by default as mentioned, it would be nice to update the corresponding manual at https://pdal.io/stages/filters.normal.html

Thanks!

**Comment by chambbj** [comment]
It is still available and the documentation does not suggest that refinement is enabled.

If you feel there is still an issue, please create a new ticket (https://github.com/PDAL/PDAL/issues/new/choose) with additional information, as what you've provided here would not be sufficient for anybody to track it down.

**Pull request merged**